multilep/src/GenAnalyzer.cc

```diff
@@ -105,8 +125,84 @@ void GenAnalyzer::analyze(const edm::Event& iEvent){
                 _gen_phPassParentage[_gen_nPh] = GenTools::passParentage(p, *genParticles);
                 ++_gen_nPh;
             } 
-        }
-    }
+	}
+
+       //store all generator level particles
+       if( multilepAnalyzer->storeGenParticles )
+	 {	    
+	    float ptGen = p.pt();
+	    float etaGen = p.eta();
+	    float phiGen = p.phi();
+	    float EGen = p.energy();
+	    int idGen = p.pdgId();
+	    int statusGen = p.status();
+	    int chargeGen = p.charge();
+	    bool isPromptFinalStateGen = p.isPromptFinalState();
+	    bool isDirectPromptTauDecayProductFinalState = p.isDirectPromptTauDecayProductFinalState();
+	    bool isLastCopy = p.isLastCopy();
+	    int indexGen = _gen_n;
+
+	    if( _gen_n == gen_n_max )
+	      {		      
+		 std::cout << "Reached max" << std::endl;
+		 exit(1);
+	      }
+       
+	    const reco::GenParticle* mom = GenTools::getMother(p, *genParticles);
+       
+	    int mother_index = 0;
+	    if( mom )
+	      {		 
+		 reco::GenParticleCollection genParticles_m = *genParticles;
+		 for(reco::GenParticleCollection::const_iterator pp = genParticles_m.begin(); pp != genParticles_m.end(); pp++)
+		   {
+		      if( fabs(pp->pt()-mom->pt()) < 10E-6 && fabs(pp->eta()-mom->eta()) < 10E-6 ) break;
+		      mother_index++;
+		   }
+	      }
+	    else mother_index = -1;
```

Comment:
I guess this code will work, but maybe it is a bit of an detour :-) Imagine you have a version of GenTools::getMother (let's call it getMotherIndex for example) which returns gen.motherRef(0).key() instead of mom in this line: https://github.com/GhentAnalysis/heavyNeutrino/blob/master/multilep/src/GenTools.cc#L16 Wouldn't that give the same result, but without the additional loop and pt/eta matching?

Reply:
Also, remember the getMother always gives the first mother, there are actually many cases in the genparticles where there will be a second mother. Maybe they might be useful too...

Reply:
Also to nitpick like crazy : ++mother_index is faster than mother_index++

Reply:
Another small remark: note that the getMother function iterates up to the next mother when the pdgId is the same. But as you are saving all gen particles for your purpose, you might as well simply save the direct mother, as you can do this iteration later in the analysis code itself.